### PR TITLE
wip multiexpression

### PR DESCRIPTION
**Pull request by bdon**
@nvkelso this is a WIP on a refactor of landuse into an expression

### CODE REVIEW

File: tiles/src/main/java/com/protomaps/basemap/layers/Landuse.java [hunk 1]
```diff
@@ -1,178 +1,109 @@
...
-            sf.hasTag("designation", "national_park") ||
-            sf.hasTag("protection_title", "National Park"))) {
-          kind = "national_park";
-        } else {
-          kind = "park";
-        }
-      }
+  // TODO craft and historic
+  public static MultiExpression.Index<String> LANDUSE_KIND = compose(
+    valueEntries("aeroway", "aerodrome", "runway"),
```
Comment (bdon) [inline tiles/src/main/java/com/protomaps/basemap/layers/Landuse.java:32]: `valueEntries` means that it should detect the key `aeroway` and return the kind as `aerodrome` or `runway` if one of those match

File: tiles/src/main/java/com/protomaps/basemap/layers/Landuse.java [hunk 2]
```diff
@@ -1,178 +1,109 @@
...
+    valueEntries("area:aeroway", "taxiway", "runway"),
+    valueEntries("amenity", "university", "college", "hospital", "library", "school"), // townhall? post_office?
+    Stream.of(entry("pedestrian", and(
```
Comment (bdon) [inline tiles/src/main/java/com/protomaps/basemap/layers/Landuse.java:35]: This means assign the kind `pedestrian` if (area = yes AND (highway = pedestrian OR highway = footway))

File: tiles/src/main/java/com/protomaps/basemap/layers/Landuse.java [hunk 3]
```diff
@@ -1,178 +1,109 @@
...
+      matchAny("area", "yes"),
+      matchAny("highway", "pedestrian", "footway")
+    ))),
+    valueEntries("landuse", "cemetery"),
+    Stream.of(entry("farmland", matchAny("landuse", "orchard", "farmland", "farmyard"))),
```
Comment (bdon) [inline tiles/src/main/java/com/protomaps/basemap/layers/Landuse.java:40]: this means assign the kind `farmland` if landuse matches any of those 3 values

File: tiles/src/main/java/com/protomaps/basemap/layers/Landuse.java [hunk 4]
```diff
@@ -1,178 +1,109 @@
...
-      //OsmNames.setOsmNames(poly, sf, 0);
-    }
+  static MatchAny PROTECTION_TITLE =
+    matchAny("protection_title", "Conservation Area", "Conservation Park", "Environmental use", "Forest Reserve",
+      "National Forest", "National Wildlife Refuge", "Nature Refuge", "Nature Reserve", "Protected Site",
+      "Provincial Park", "Public Access Land", "Regional Reserve", "Resources Reserve", "State Forest",
+      "State Game Land", "State Park", "Watershed Recreation Unit", "Wild Forest", "Wilderness Area",
+      "Wilderness Study Area", "Wildlife Management", "Wildlife Management Area", "Wildlife Sanctuary");
+
+//  public static Stream<MultiExpression.Entry<String>> US_LANDUSE_KIND = compose(
```
Comment (bdon) [inline tiles/src/main/java/com/protomaps/basemap/layers/Landuse.java:71]: not implemented yet...

File: tiles/src/main/java/com/protomaps/basemap/layers/Landuse.java [hunk 5]
```diff
@@ -1,178 +1,109 @@
...
+    valueEntries("landuse", "residential"),
+    Stream.of(entry("industrial", matchAny("landuse", "industrial", "brownfield"))),
+    valueEntries("landuse", "military"),
+    valueEntries("military", "naval_base", "airfield"),
+    valueEntries("leisure", "golf_course", "park", "stadium", "garden", "dog_park", "playground", "pitch",
+      "nature_reserve"),
+    Stream.of(entry("pedestrian", matchAny("man_made", "bridge"))),
+    valueEntries("man_made", "pier"),
+    valueEntries("natural", "beach"),
+    valueEntries("shop", "grocery", "supermarket"), // not in tilezen?
```
Comment (nvkelso) [inline tiles/src/main/java/com/protomaps/basemap/layers/Landuse.java:50]: @nvkelso to research

File: tiles/src/test/java/com/protomaps/basemap/layers/LanduseTest.java [hunk 6]
```diff
@@ -21,4 +23,64 @@ void simple() {
...
+    assertKind("farmland", Map.of("landuse", "farmyard"));
+    assertKind("residential", Map.of("landuse", "residential"));
+    assertKind("industrial", Map.of("landuse", "industrial"));
+    assertKind("industrial", Map.of("landuse", "brownfield"));
+    assertKind("military", Map.of("landuse", "military"));
+    //    assertKind("naval_base", Map.of("military", "naval_base"));
+    //    assertKind("airfield", Map.of("military", "airfield"));
+    assertKind("golf_course", Map.of("leisure", "golf_course"));
+    //    assertKind("marina", Map.of("leisure", "marina"));
+    assertKind("park", Map.of("leisure", "park"));
```
Comment (nvkelso) [inline tiles/src/test/java/com/protomaps/basemap/layers/LanduseTest.java:76]: To test the order of operations, there should at least be one test that includes other tags than just the expected minimum tags... eg add a boundary key-value in.

**Comment by bdon** [comment]
* should fix `area:aeroway` bug with missing kind
* fix amenity=post_office, amenity=townhall

**Pull request closed without merging**